### ISSUE TITLE
Emacs freezes when typing Rmd files

### ISSUE THREAD

**Issue by fmichonneau** [vspinu/polymode#87]
I don't know if it's related to https://stat.ethz.ch/pipermail/ess-help/2016-January/010876.html but with the latest melpa version of polymode (20160215.1711), when typing in an Rmd file, I have regular freezes of Emacs (that then needs to be killed). Sometimes the freeze is concordant with the following error message (but other times, the error message doesn't cause the freeze)...

`(poly-lock-fontify region 2842 2844) -> (font-lock-default-fontify-region 2842 2844): Args out of range 2803, 2803`

**Comment by vspinu** [comment]
Could you please send a sample of the file? 

What version of emacs are you using? 

There is an emacs bug that was causing it. I added some workarounds.

These sort of bugs are easy to fix, I just need to be able to reproduce those.

**Comment by vspinu** [comment]
I cannot reproduce this with files that I have on emacs24.4 and 25.1. Earlier versions of emacs will probably fail. `advice-add` is needed. 

Could you please provide an example file?

**Comment by vspinu** [comment]
> I don't know if it's related to https://stat.ethz.ch/pipermail/ess-help/2016-January/010876.html 

It's probably unrelated. Current polymode is a complete rewrite with many such issues fixed.

**Comment by gragusa** [comment]
I get similar freeze when using `polymode` to edit `Rnw` files using Emacs 24.9. I was getting a lot of `font-lock-default-fontify-region` `out-of-range` messages. The freezes occurred  when entering into code blocks. I will try to have a reproducible test case.

**Comment by fernandomayer** [comment]
I'm having the same issues with the latest version of polymode in Emacs 24.5.1 with Spacemacs 0.105.11.

I think it's difficult to have a reproducible example. For me the freezes occur randomly. Sometimes when entering in a block code, sometimes just typing something anywhere in the text.

When I have a YAML block code in the top of a Rmd file and I try to edit something in this block code, the freeze occurs immediately.

**Comment by tim-teufel** [comment]
Well, Spacemacs…
Could you please check if you can get the freeze [in an `emacs -Q` session](http://www.lunaryorn.com/2015/11/29/reproduce-bugs-in-emacs-Q.html)?

**Comment by vspinu** [comment]
@fernandomayer 

After C-g you  should see polymode messages in `*Messages*` buffer. Like 

```
(poly-lock-fontify-region 2842 2844) -> (font-lock-default-fontify-region 2842 2844): Args out of range 2803, 2803
```

Those are valid commands. That means that you can execute them with `M-: (poly-lock-fontify-region 2842 2844)`. If it locks just post your file here (if you can) and I will take it from there. 

The fact that this issue affects multiple users should mean that there is indeed a lurking problem somewhere.

**Comment by gragusa** [comment]
The problem is that is very difficult to reduce the problem to a test case.
It seems happening at random with the only regularity being the it freezes
after editing R chunk soon after a new chunk has being created.

**Comment by vspinu** [comment]
I can reproduce errors with the recent markdown on all emacses. There seem to have been some low level changes on how markdown fontifies stuff. I am tracking that down right now. 

@fernandomayer  Can I have an example of markdown with YAML that gave you trouble?

**Comment by fernandomayer** [comment]
@vspinu 

[This gist](https://gist.github.com/fernandomayer/65e44e5fa93236ff4bb8) have a simple example. It gave me this font-lock error when I tried to edit the YAML, but also freezed in other parts of the text too.

Note that it have an empty chunk at the end. I leave it there on purpose, because for me, *always* when I continue writing the phrase above it, it gives me the freeze. But, even without this empty chunk, the freeze occur in some part of the file.

I don't think this is a Spacemacs issue, since other users are also experiencing it, and I know other ("pure") Emacs users that are having the same problem.

**Comment by vspinu** [comment]
Thanks @fernandomayer. If you don't mind I will add this sample to the test directory in polymode. 

I have tracked the problem the other day and it turned to be on markdown side. While I can have an advice in polymode I am waiting for my fix to be merged [here](https://github.com/jrblevin/markdown-mode/pull/110). You can clone the markdown from my fork if you need it really fast or put the following in your emacs:

```elisp
(defun markdown-match-propertized-text (property last)
  "Match text with PROPERTY from point to LAST.
Restore match data previously stored in PROPERTY."
  (let ((saved (get-text-property (point) property))
        pos)
    (unless saved
      (setq pos (next-single-char-property-change (point) property nil last))
      (setq saved (get-text-property pos property)))
    (when saved
      (set-match-data saved)
      ;; Step at least one character beyond point. Otherwise
      ;; `font-lock-fontify-keywords-region' infloops.
      (goto-char (min (1+ (max (match-end 0) (point)))
                      (point-max)))
      saved)))
```

It will be actually very helpful if you could try it out and see if that fixes your freezes. Thanks.

**Comment by fernandomayer** [comment]
@vspinu no problem to add this as a sample if you find it useful.

I've included your modified function in my `markdown-mode.el`. 

I've edited this same sample file, trying to change the YAML, chunks, and text in several parts. It didn't freeze in any moment, although the `(poly-lock-fontify-region ... )` message keeps popping up at the mini-buffer. Anyway, it seems to have solved the problem.

I will see how it goes with other files, and I will let you know if something different happens.

**Comment by vspinu** [comment]
> I've included your modified function in my markdown-mode.el. 

You can simply put it into your .emacs.

> (poly-lock-fontify-region ... ) message keeps popping up 

Make sure you have the most recent markdown-mode.el. There was a small patch last week that removes a bunch of those messages.

**Comment by fernandomayer** [comment]
Thanks @vspinu, that worked.

I updated markdown-mode. The message appears only once in the beginning. I've tested in some other files I have (some very big and complex Rmd files), and it didn't freeze anymore!

**Comment by kjhealy** [comment]
Thanks so much for working to track down and fix this—I've been having this issue as well and it's been driving me crazy. I'll let you know how your fix goes for me.

**Comment by gragusa** [comment]
Yes, thank you @vspinu. I have not applied the patch yet. When I do I report back whether the freezes subsides. By the way, has anybody experienced freezes when editing Rnw files? I wars just editing one and I got a freeze (but it could be unrelated to polymode).
gr




On Mon, Feb 29, 2016 at 8:27 AM -0800, "Kieran Healy" <notifications@github.com> wrote:












Thanks so much for working to track down and fix this—I've been having this issue as well and it's been driving me crazy. I'll let you know how your fix goes for me.

**Comment by kjhealy** [comment]
@vspinu — Some followup FWIW: with your function in my `.emacs.d/`, I'm still seeing this locking behavior, and not just in `.Rmd` files. In an `.md` file with some YAML at the top it happens repeatedly within a few seconds of trying to edit the file. E.g., loading this gist as a file— https://gist.github.com/kjhealy/fcc22b29dbcd9a72bfc2—reliably causes my Emacs to lock-up with a few seconds of trying to edit or move around the file. Doing `M-x markdown-mode` (to turn off polymode) seems to make the problem go away. In `.Rmd` files the problem seems to take longer to trigger, as reported (and as I was experiencing myself).

I know this is still a mostly anecdotal report, as I haven't had a chance to set up a clean emacs that loads just polymode/markdown-mode yet. Thanks again for your work on this.

**Comment by vspinu** [comment]
Thanks @kjhealy. I will have a look. What version of emacs is this?

**Comment by kjhealy** [comment]
I can make both 24.5 and the 25.0.91 pretest lock up in this way.

**Comment by vspinu** [comment]
I see some funny font-locking with markdown headers but no locking. Could you please do the following. Start new emacs session and execute:

```
(pm-debug-trace-functions-by-regexp "font-lock-")
(pm-debug-trace-functions-by-regexp "poly-lock-")
(pm-debug-trace-functions-by-regexp "markdown")
(pm-debug-trace-functions-by-regexp "polymode-.*function")
(pm-debug-trace-relevant-functions "insert")
(pm-debug-trace-functions-by-regexp "^pm-")
```

Then go to .md file and start typing. When lock occurs wait for a couple of seconds then unlock with C-g and move to *trace-output* buffer. Then save that buffer and send it to me or post it on gist. This way I can have a clear picture of the location when the infloop occurs.

**Comment by vspinu** [comment]
Also please make sure that you are using the up to date version of polymode. I am continuously making changes these days, so a matter of  a couple of hours can make quite a difference.

**Comment by kjhealy** [comment]
Hm, updating to the latest polymode (I was only a day or two behind I think) seems to have changed things for the better. I'll keep trying to find a solidly reproducible case.

**Comment by vspinu** [comment]
The freezes are definitely gone with most recent polymode and markdown-mode. There are still occasional out of range errors but those should not impact the experience too much. I am looking into how to completely get rid of those.

**Comment by bbolker** [comment]
I am getting superficially similar freezes with what I think are the current versions of polymode/markdown-mode. If anyone has a simple incantation for retrieving my polymode/markdown-mode versions so I can report it that would be greatly appreciated (this is on Emacs 24.3.1). Roughly speaking, the freezes occur as I start to create a new R-mode chunk. (Happy to move this to a new issue if that would be more useful.)
